dsp/templates/template_v2.py

```diff
@@ -142,6 +142,11 @@ def extract(
         """
         example = dsp.Example(example)
 
+        # Unset the output variables to prevent leakage of gold answers
+        for field in self.fields:
+            if field.output_variable in example:
```

Comment:
Thanks a lot for opening this PR! Q: Isn't this a bit too aggressive? My quick impression is that it'll basically unset *all* values in the example

Reply:
Thanks for the review! I have done the following changes:

- Now we only unset a part of fields. In `do_generate` and `extract` function, I've added an optional argument called `output_fields`, meaning the fields that user expect the LM to output/extract. The default value fo `output_fields` in `do_generate` is `['answer']`, which I believe is the most common use case. User can specify the output fields as well. For example, if the user expects to extract the `rationale`, it can be done by
```python
dsp.generate(my_qa_template_with_passages)(example, stage='qa', output_fields=['answer', 'rationale'])
```

- As you requested, I added some unit tests to illustrate how `extract` function behave after this change. Test cases includes some well-formed raw reponses, and a malformed one. You can run them using
```
python -m unittest tests.extract_test
```